### ISSUE TITLE
FileNotFoundException when using custom serializer

### ISSUE THREAD

**Issue by christofferjerrebo** [Microsoft/service-fabric-services-and-actors-dotnet#147]
We recently upgraded from 3.1.301 to 3.2.176 and noticed that the Microsoft.ServiceFabric.Services.Remoting.V2 interfaces have changed. We changed our custom serializer following this guide: https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-reliable-services-communication-remoting#how-to-use-custom-serialization-with-remoting-wrapped-message

However, now we are getting the following exception:
```csharp
"One or more errors occurred. (Error resolving type specified in JSON 'Security.Api.IUserService_.service.mt.GetUsersAsyncReqBody, Security.Api.IUserService_.service.mt'. Path 'value.$type', line 1, position 193.)"
```
With InnerException:
```csharp
{System.IO.FileNotFoundException: Could not load file or assembly 'Security.Api.IUserService_.service.mt, Culture=neutral, PublicKeyToken=null'. The system cannot find the file specified.
File name: 'Security.Api.IUserService_.service.mt, Culture=neutral, PublicKeyToken=null'
   at System.Reflection.RuntimeAssembly.nLoad(AssemblyName fileName, String codeBase, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, IntPtr ptrLoadContextBinder)
   at System.Reflection.RuntimeAssembly.InternalLoadAssemblyName(AssemblyName assemblyRef, RuntimeAssembly reqAssembly, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, IntPtr ptrLoadContextBinder)
   at System.Reflection.RuntimeAssembly.InternalLoad(String assemblyString, StackCrawlMark& stackMark, IntPtr pPrivHostBinder)
   at System.Reflection.RuntimeAssembly.InternalLoad(String assemblyString, StackCrawlMark& stackMark)
   at System.Reflection.Assembly.Load(String assemblyString)
   at System.Reflection.Assembly.LoadWithPartialName(String partialName)
   at Newtonsoft.Json.Serialization.DefaultSerializationBinder.GetTypeFromTypeNameKey(TypeNameKey typeNameKey)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at Newtonsoft.Json.Serialization.DefaultSerializationBinder.BindToType(String assemblyName, String typeName)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.ResolveTypeName(JsonReader reader, Type& objectType, JsonContract& contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, String qualifiedTypeName)
}
```

From what we can understand, the deserialization can't find the type. However, the type seems to be generated on the fly considering the `_.service.mt`. Can someone point us in the right direction?

**Comment by amanbha** [comment]
@christofferjerrebo Could you please clarify if you started seeing this exception after changing your serializer when you upgraded to 3.2.176 and it was working fine before it? Would it be possible to point to changed code?

**Comment by christofferjerrebo** [comment]
@amanbha Yes we started seeing this exception after upgrade to 3.2.176 and changing the serializer. Before this, the serialization worked fine. I'll include the commit diff, hope this helps.

```csharp
diff --git a/src/Xyz.Services.Remoting/ServiceFabricJsonSerializer/JsonBody.cs b/src/Xyz.Services.Remoting/ServiceFabricJsonSerializer/JsonBody.cs
new file mode 100644
index 0000000..1c2963d
--- /dev/null
+++ b/src/Xyz.Services.Remoting/ServiceFabricJsonSerializer/JsonBody.cs
@@ -0,0 +1,34 @@
+namespace Xyz.Services.Remoting
+{
+    using System;
+    using Microsoft.ServiceFabric.Services.Remoting.V2;
+
+    internal sealed class JsonBody : WrappedMessage, IServiceRemotingRequestMessageBody, IServiceRemotingResponseMessageBody
+    {
+        public JsonBody(object wrapped)
+        {
+            this.Value = wrapped;
+        }
+
+        public void SetParameter(int position, string parameName, object parameter)
+        {  //Not Needed if you are using WrappedMessage
+            throw new NotImplementedException();
+        }
+
+        public object GetParameter(int position, string parameName, Type paramType)
+        {
+            //Not Needed if you are using WrappedMessage
+            throw new NotImplementedException();
+        }
+
+        public void Set(object response)
+        { //Not Needed if you are using WrappedMessage
+            throw new NotImplementedException();
+        }
+
+        public object Get(Type paramType)
+        {  //Not Needed if you are using WrappedMessage
+            throw new NotImplementedException();
+        }
+    }
+}
\ No newline at end of file
diff --git a/src/Xyz.Services.Remoting/ServiceFabricJsonSerializer/JsonMessageFactory.cs b/src/Xyz.Services.Remoting/ServiceFabricJsonSerializer/JsonMessageFactory.cs
index cd8b6a5..0204180 100644
--- a/src/Xyz.Services.Remoting/ServiceFabricJsonSerializer/JsonMessageFactory.cs
+++ b/src/Xyz.Services.Remoting/ServiceFabricJsonSerializer/JsonMessageFactory.cs
@@ -4,15 +4,14 @@
 
     internal sealed class JsonMessageFactory : IServiceRemotingMessageBodyFactory
     {
-        public IServiceRemotingRequestMessageBody CreateRequest(string interfaceName, string methodName,
-            int numberOfParameters)
+        public IServiceRemotingRequestMessageBody CreateRequest(string interfaceName, string methodName, int numberOfParameters, object wrappedRequestObject)
         {
-            return new JsonRemotingRequestBody();
+            return new JsonBody(wrappedRequestObject);
         }
 
-        public IServiceRemotingResponseMessageBody CreateResponse(string interfaceName, string methodName)
+        public IServiceRemotingResponseMessageBody CreateResponse(string interfaceName, string methodName, object wrappedRequestObject)
         {
-            return new JsonRemotingResponseBody();
+            return new JsonBody(wrappedRequestObject);
         }
     }
 }
\ No newline at end of file
diff --git a/src/Xyz.Services.Remoting/ServiceFabricJsonSerializer/JsonRemotingRequestBody.cs b/src/Xyz.Services.Remoting/ServiceFabricJsonSerializer/JsonRemotingRequestBody.cs
deleted file mode 100644
index b6538ea..0000000
--- a/src/Xyz.Services.Remoting/ServiceFabricJsonSerializer/JsonRemotingRequestBody.cs
+++ /dev/null
@@ -1,21 +0,0 @@
-﻿namespace Xyz.Services.Remoting
-{
-    using System;
-    using System.Collections.Generic;
-    using Microsoft.ServiceFabric.Services.Remoting.V2;
-
-    internal class JsonRemotingRequestBody : IServiceRemotingRequestMessageBody
-    {
-        public readonly Dictionary<string, object> Parameters = new Dictionary<string, object>();
-
-        public void SetParameter(int position, string parameName, object parameter)
-        {
-            Parameters[parameName] = parameter;
-        }
-
-        public object GetParameter(int position, string parameName, Type paramType)
-        {
-            return Parameters[parameName];
-        }
-    }
-}
\ No newline at end of file
diff --git a/src/Xyz.Services.Remoting/ServiceFabricJsonSerializer/JsonRemotingResponseBody.cs b/src/Xyz.Services.Remoting/ServiceFabricJsonSerializer/JsonRemotingResponseBody.cs
deleted file mode 100644
index a93d5b6..0000000
--- a/src/Xyz.Services.Remoting/ServiceFabricJsonSerializer/JsonRemotingResponseBody.cs
+++ /dev/null
@@ -1,20 +0,0 @@
-﻿namespace Xyz.Services.Remoting
-{
-    using System;
-    using Microsoft.ServiceFabric.Services.Remoting.V2;
-
-    internal class JsonRemotingResponseBody : IServiceRemotingResponseMessageBody
-    {
-        public object Value;
-
-        public void Set(object response)
-        {
-            Value = response;
-        }
-
-        public object Get(Type paramType)
-        {
-            return Value;
-        }
-    }
-}
\ No newline at end of file
diff --git a/src/Xyz.Services.Remoting/ServiceFabricJsonSerializer/ServiceRemotingJsonSerializationProvider.cs b/src/Xyz.Services.Remoting/ServiceFabricJsonSerializer/ServiceRemotingJsonSerializationProvider.cs
index 677c77e..1e73fc1 100644
--- a/src/Xyz.Services.Remoting/ServiceFabricJsonSerializer/ServiceRemotingJsonSerializationProvider.cs
+++ b/src/Xyz.Services.Remoting/ServiceFabricJsonSerializer/ServiceRemotingJsonSerializationProvider.cs
@@ -6,21 +6,19 @@
 
     public sealed class ServiceRemotingJsonSerializationProvider : IServiceRemotingMessageSerializationProvider
     {
-        public IServiceRemotingRequestMessageBodySerializer CreateRequestMessageSerializer(Type serviceInterfaceType,
-            IEnumerable<Type> requestBodyTypes)
+        public IServiceRemotingMessageBodyFactory CreateMessageBodyFactory()
         {
-            return new ServiceRemotingRequestJsonMessageBodySerializer(serviceInterfaceType, requestBodyTypes);
+            return new JsonMessageFactory();
         }
 
-        public IServiceRemotingResponseMessageBodySerializer CreateResponseMessageSerializer(Type serviceInterfaceType,
-            IEnumerable<Type> responseBodyTypes)
+        public IServiceRemotingRequestMessageBodySerializer CreateRequestMessageSerializer(Type serviceInterfaceType, IEnumerable<Type> requestWrappedType, IEnumerable<Type> requestBodyTypes = null)
         {
-            return new ServiceRemotingResponseJsonMessageBodySerializer(serviceInterfaceType, responseBodyTypes);
+            return new ServiceRemotingRequestJsonMessageBodySerializer();
         }
 
-        public IServiceRemotingMessageBodyFactory CreateMessageBodyFactory()
+        public IServiceRemotingResponseMessageBodySerializer CreateResponseMessageSerializer(Type serviceInterfaceType, IEnumerable<Type> responseWrappedType, IEnumerable<Type> responseBodyTypes = null)
         {
-            return new JsonMessageFactory();
+            return new ServiceRemotingResponseJsonMessageBodySerializer();
         }
     }
 }
\ No newline at end of file
diff --git a/src/Xyz.Services.Remoting/ServiceFabricJsonSerializer/ServiceRemotingRequestJsonMessageBodySerializer.cs b/src/Xyz.Services.Remoting/ServiceFabricJsonSerializer/ServiceRemotingRequestJsonMessageBodySerializer.cs
index 231d13a..6241173 100644
--- a/src/Xyz.Services.Remoting/ServiceFabricJsonSerializer/ServiceRemotingRequestJsonMessageBodySerializer.cs
+++ b/src/Xyz.Services.Remoting/ServiceFabricJsonSerializer/ServiceRemotingRequestJsonMessageBodySerializer.cs
@@ -7,47 +7,43 @@
     using Microsoft.ServiceFabric.Services.Remoting.V2.Messaging;
     using Newtonsoft.Json;
 
-    internal class ServiceRemotingRequestJsonMessageBodySerializer : IServiceRemotingRequestMessageBodySerializer
+    internal sealed class ServiceRemotingRequestJsonMessageBodySerializer : IServiceRemotingRequestMessageBodySerializer
     {
-        public ServiceRemotingRequestJsonMessageBodySerializer(Type serviceInterfaceType,
-            IEnumerable<Type> parameterInfo)
+        private readonly JsonSerializer _serializer;
+
+        public ServiceRemotingRequestJsonMessageBodySerializer()
         {
+            _serializer = JsonSerializer.Create(new JsonSerializerSettings()
+            {
+                TypeNameHandling = TypeNameHandling.All
+            });
         }
 
-        public OutgoingMessageBody Serialize(IServiceRemotingRequestMessageBody serviceRemotingRequestMessageBody)
+        public IOutgoingMessageBody Serialize(IServiceRemotingRequestMessageBody serviceRemotingRequestMessageBody)
         {
             if (serviceRemotingRequestMessageBody == null)
             {
                 return null;
             }
-
             using (var writeStream = new MemoryStream())
             using (var jsonWriter = new JsonTextWriter(new StreamWriter(writeStream)))
             {
-                var serializer = JsonSerializer.Create(new JsonSerializerSettings
-                {
-                    TypeNameHandling = TypeNameHandling.All
-                });
-                serializer.Serialize(jsonWriter, serviceRemotingRequestMessageBody);
-
+                _serializer.Serialize(jsonWriter, serviceRemotingRequestMessageBody);
                 jsonWriter.Flush();
-                var segment = new ArraySegment<byte>(writeStream.ToArray());
-                var segments = new List<ArraySegment<byte>> {segment};
+                var bytes = writeStream.ToArray();
+                var segment = new ArraySegment<byte>(bytes);
+                var segments = new List<ArraySegment<byte>> { segment };
                 return new OutgoingMessageBody(segments);
             }
         }
 
-        public IServiceRemotingRequestMessageBody Deserialize(IncomingMessageBody messageBody)
+        public IServiceRemotingRequestMessageBody Deserialize(IIncomingMessageBody messageBody)
         {
             using (var sr = new StreamReader(messageBody.GetReceivedBuffer()))
-            using (var reader = new JsonTextReader(sr))
+            using (JsonReader reader = new JsonTextReader(sr))
             {
-                var serializer = JsonSerializer.Create(new JsonSerializerSettings
-                {
-                    TypeNameHandling = TypeNameHandling.All
-                });
-
-                return serializer.Deserialize<JsonRemotingRequestBody>(reader);
+                var ob = _serializer.Deserialize<JsonBody>(reader);
+                return ob;
             }
         }
     }
diff --git a/src/Xyz.Services.Remoting/ServiceFabricJsonSerializer/ServiceRemotingResponseJsonMessageBodySerializer.cs b/src/Xyz.Services.Remoting/ServiceFabricJsonSerializer/ServiceRemotingResponseJsonMessageBodySerializer.cs
index 24ad5f2..8065df8 100644
--- a/src/Xyz.Services.Remoting/ServiceFabricJsonSerializer/ServiceRemotingResponseJsonMessageBodySerializer.cs
+++ b/src/Xyz.Services.Remoting/ServiceFabricJsonSerializer/ServiceRemotingResponseJsonMessageBodySerializer.cs
@@ -8,36 +8,44 @@
     using Microsoft.ServiceFabric.Services.Remoting.V2.Messaging;
     using Newtonsoft.Json;
 
-    internal class ServiceRemotingResponseJsonMessageBodySerializer : IServiceRemotingResponseMessageBodySerializer
+    internal sealed class ServiceRemotingResponseJsonMessageBodySerializer : IServiceRemotingResponseMessageBodySerializer
     {
-        public ServiceRemotingResponseJsonMessageBodySerializer(Type serviceInterfaceType,
-            IEnumerable<Type> parameterInfo)
-        {
-        }
+        private readonly JsonSerializer _serializer;
 
-        public OutgoingMessageBody Serialize(IServiceRemotingResponseMessageBody responseMessageBody)
+        public ServiceRemotingResponseJsonMessageBodySerializer()
         {
-            var json = JsonConvert.SerializeObject(responseMessageBody, new JsonSerializerSettings
+            _serializer = JsonSerializer.Create(new JsonSerializerSettings()
             {
                 TypeNameHandling = TypeNameHandling.All
             });
-            var bytes = Encoding.UTF8.GetBytes(json);
-            var segment = new ArraySegment<byte>(bytes);
-            var list = new List<ArraySegment<byte>> {segment};
-            return new OutgoingMessageBody(list);
         }
 
-        public IServiceRemotingResponseMessageBody Deserialize(IncomingMessageBody messageBody)
+        public IOutgoingMessageBody Serialize(IServiceRemotingResponseMessageBody responseMessageBody)
+        {
+            if (responseMessageBody == null)
+            {
+                return null;
+            }
+
+            using (var writeStream = new MemoryStream())
+            using (var jsonWriter = new JsonTextWriter(new StreamWriter(writeStream)))
+            {
+                _serializer.Serialize(jsonWriter, responseMessageBody);
+                jsonWriter.Flush();
+                var bytes = writeStream.ToArray();
+                var segment = new ArraySegment<byte>(bytes);
+                var segments = new List<ArraySegment<byte>> { segment };
+                return new OutgoingMessageBody(segments);
+            }
+        }
+
+        public IServiceRemotingResponseMessageBody Deserialize(IIncomingMessageBody messageBody)
         {
             using (var sr = new StreamReader(messageBody.GetReceivedBuffer()))
             using (var reader = new JsonTextReader(sr))
             {
-                var serializer = JsonSerializer.Create(new JsonSerializerSettings
-                {
-                    TypeNameHandling = TypeNameHandling.All
-                });
-
-                return serializer.Deserialize<JsonRemotingResponseBody>(reader);
+                var obj = _serializer.Deserialize<JsonBody>(reader);
+                return obj;
             }
         }
     }
```

**Comment by christofferjerrebo** [comment]
The `ServiceInstanceListener` for each service created with:
```csharp
return new ServiceInstanceListener(serviceContext =>
{
    // Custom JSON serializer for handling immutable collections
    var serializationProvider = new ServiceRemotingJsonSerializationProvider();

    // Custom service remote message handler for extracting the request headers and adding them as properties to the log.
    var serviceRemotingMessageHandler = new LogHeaderServiceRemotingMessageHandler(serviceContext, service,
        serializationProvider.CreateMessageBodyFactory());

    var remotingListenerSettings = new FabricTransportRemotingListenerSettings { UseWrappedMessage = true };

    // Create service instance listener
    return new FabricTransportServiceRemotingListener(
        serviceContext,
        serviceRemotingMessageHandler,
        remotingListenerSettings,
        serializationProvider);
});
```
The `ServiceProxyFactory` for our Web API gateway is created using:
```csharp
private void AddServices(IServiceCollection services)
{
    services.AddSingleton<IServiceProxyFactory>(provider =>
        new ServiceProxyFactory(c =>
            new ServiceRemotingClientFactoryWrapper(
                provider.GetRequiredService<IHttpContextAccessor>(),
                new FabricTransportServiceRemotingClientFactory(
                    remotingSettings: new Microsoft.ServiceFabric.Services.Remoting.FabricTransport.FabricTransportRemotingSettings { UseWrappedMessage = true },
                    serializationProvider: new ServiceRemotingJsonSerializationProvider())
            )));

// Removed some code

}
```

**Comment by VipulM-MSFT** [comment]
The wrapped messages allows the interface compatibility with the V1 remoting stack. If you had object the parameter, the unwrapped messages would not work.

**Comment by christofferjerrebo** [comment]
Since we don't need interface compatibility I switched off wrapped messages. The serialization is more or less the same as previous. It works now, closing this issue.

**Comment by rfcdejong** [comment]
@christofferjerrebo We run into the same error as you, how did you switched off wrapped messages?

**Comment by christofferjerrebo** [comment]
Basically we used the old implementation of the service remoting. Using the "Custom serialization" implementation from https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-reliable-services-communication-remoting#use-custom-serialization-with-a-remoting-wrapped-message, the changes are on `JsonMessageFactory` and when creating a `ServiceProxyFactory`.

```charp
internal sealed class JsonMessageFactory : IServiceRemotingMessageBodyFactory
{
    public IServiceRemotingRequestMessageBody CreateRequest(string interfaceName, string methodName, int numberOfParameters, object wrappedRequestObject) =>
        new JsonRemotingRequestBody();

    public IServiceRemotingResponseMessageBody CreateResponse(string interfaceName, string methodName, object wrappedResponseObject) =>
        new JsonRemotingResponseBody();
}

internal sealed class JsonRemotingRequestBody : IServiceRemotingRequestMessageBody
{
    public readonly Dictionary<string, object> Parameters = new Dictionary<string, object>();

    public void SetParameter(int position, string parameName, object parameter) =>
        Parameters[parameName] = parameter;

    public object GetParameter(int position, string parameName, Type paramType) =>
        Parameters[parameName];
}

internal sealed class JsonRemotingResponseBody : IServiceRemotingResponseMessageBody
{
    public object Value;

    public void Set(object response) =>
        Value = response;

    public object Get(Type paramType) =>
        Value;
}
```

Make sure you remove the `settings.UseWrappedMessage = true;` line when creating a `ServiceProxyFactory`.

**Comment by rfcdejong** [comment]
@christofferjerrebo We had many issues serializing over Json, the wrapped functionality of Service Fabric sample is simply not working when type information is required. We ended up writing a lot of custom stuff for:

- SerializationBinder (based on https://github.com/Microsoft/Bing-Maps-Fleet-Tracker/blob/master/Backend/src/Trackable.Func/Shared/AdvancedSerializationBinder.cs from a Microsoft employee working at Bing Maps, but I had to fix a RegEx pattern matching first)

- ContractResolver (based on https://stackoverflow.com/questions/38336390/deserialize-dictionarystring-object-with-enum-values-in-c-sharp/38340375#38340375)

- Several JsonConverters
-> For creating an ActorId object
-> For custom type resolving where we use runtime objects (dynamic) which are not static defined at all.

@suchiagicha Doesn't work either way, I cannot reproduce the _service.mt exceptions anymore, but now it keeps complaining about $value not being able to serialize.
